vault/resource_database_secrets_mount.go

```diff
@@ -0,0 +1,334 @@
+	} else {
+		cur = append(cur, vals)
+	}
+	s.result[db.Name()] = cur
+}
+
+func (s *dbConfigStore) Get(db *dbEngine) []map[string]interface{} {
+	s.m.RLock()
+	defer s.m.RUnlock()
+	if s.result == nil {
+		return nil
+	}
+
+	return s.result[db.Name()]
+}
+
+func (s *dbConfigStore) Result() map[string][]map[string]interface{} {
+	s.m.RLock()
+	defer s.m.RUnlock()
+	if s.result == nil {
+		return make(map[string][]map[string]interface{})
+	}
+
+	return s.result
+}
+
+func (s *dbConfigStore) init() {
+	if s.result == nil {
+		s.result = make(map[string][]map[string]interface{})
+	}
+}
+
+func databaseSecretsMountCustomizeDiff(ctx context.Context, d *schema.ResourceDiff, meta interface{}) error {
+	// compute the number of configured database engines
+	var engineCount int
+	for _, engine := range dbEngines {
+		count := d.Get(fmt.Sprintf("%s.#", engine)).(int)
+		if count > 0 {
+			for i := 0; i < count; i++ {
+				key := fmt.Sprintf("%s.%d.name", engine, i)
+				o, n := d.GetChange(key)
+				// don't force new on engine addition
+				if o.(string) != "" && o.(string) != n.(string) {
+					if err := d.ForceNew(key); err != nil {
+						return err
+					}
+				}
+			}
+		}
+		engineCount += count
+	}
+
+	key := "engine_count"
+	curCount := d.Get(key).(int)
+	// set the new engine count
+	if err := d.SetNew(key, engineCount); err != nil {
+		return err
+	}
+
+	if engineCount < curCount {
+		// force new resource creation if the number of configured engines has
+		// decreased.
+		// Once VAULT-5302 is fixed we can optionally apply the
+		// ForceNew if the target Vault version has the fix.
+		if err := d.ForceNew(key); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func databaseSecretsMountResource() *schema.Resource {
+	return &schema.Resource{
+		Create:        databaseSecretsMountCreateOrUpdate,
+		Read:          databaseSecretsMountRead,
+		Update:        databaseSecretsMountCreateOrUpdate,
+		Delete:        databaseSecretsMountDelete,
+		CustomizeDiff: databaseSecretsMountCustomizeDiff,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+		Schema: getDatabaseSecretsMountSchema(),
+	}
+}
+
+func getDatabaseSecretsMountSchema() schemaMap {
+	s := getMountSchema("type")
+	for k, v := range getDatabaseSchema(schema.TypeList) {
+		v.ConflictsWith = nil
+		v.MaxItems = 0
+		addCommonDatabaseSchema(v)
+		s[k] = v
+	}
+
+	// Used to gauge whether the resource should be recreated via ForceNew.
+	// If the number of engines is reduced then the resource will be recreated.
+	// This is handled in the Resource's CustomizeDiff function.
+	s["engine_count"] = &schema.Schema{
+		Type:        schema.TypeInt,
+		Computed:    true,
+		Description: "Total number of database secret engines configured under the mount.",
+	}
+
+	return s
+}
+
+func addCommonDatabaseSchema(s *schema.Schema) {
+	elem := s.Elem.(*schema.Resource)
+	for k, v := range getCommonDatabaseSchema() {
+		// TODO handle intersection errors
```

Comment:
Is this still a concern?

Reply:
Not a huge concern IMO, I think we can punt on it for now.